builtin/logical/pki/path_fetch_keys.go

```diff
@@ -0,0 +1,225 @@
+the certificate.
+`
+)
+
+func (b *backend) pathGetKeyHandler(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	keyRef := data.Get("ref").(string)
+	if len(keyRef) == 0 {
+		return logical.ErrorResponse("missing key reference"), nil
+	}
+
+	keyId, err := resolveKeyReference(ctx, req.Storage, keyRef)
+	if err != nil {
+		return nil, err
+	}
+	if keyId == "" {
+		return logical.ErrorResponse("unable to resolve key id for reference" + keyRef), nil
+	}
+
+	key, err := fetchKeyById(ctx, req.Storage, keyId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			keyIdParam:   key.ID,
+			keyNameParam: key.Name,
+			keyTypeParam: key.PrivateKeyType,
+		},
+	}, nil
+}
+
+func (b *backend) pathUpdateKeyHandler(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	keyRef := data.Get(keyRefParam).(string)
+	if len(keyRef) == 0 {
+		return logical.ErrorResponse("missing key reference"), nil
+	}
+
+	keyId, err := resolveKeyReference(ctx, req.Storage, keyRef)
+	if err != nil {
+		return nil, err
+	}
+	if keyId == "" {
+		return logical.ErrorResponse("unable to resolve key id for reference" + keyRef), nil
+	}
+
+	key, err := fetchKeyById(ctx, req.Storage, keyId)
+	if err != nil {
+		return nil, err
+	}
+
+	newName := data.Get(keyNameParam).(string)
+	if len(newName) > 0 && !nameMatcher.MatchString(newName) {
+		return logical.ErrorResponse("new key name outside of valid character limits"), nil
+	}
+
+	if newName != key.Name {
+		key.Name = newName
+
+		err := writeKey(ctx, req.Storage, *key)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			keyIdParam:   key.ID,
+			keyNameParam: key.Name,
+			keyTypeParam: key.PrivateKeyType,
+		},
+	}
+
+	if len(newName) == 0 {
+		resp.AddWarning("Name successfully deleted, you will now need to reference this key by it's Id: " + string(key.ID))
+	}
+
+	return resp, nil
+}
+
+func (b *backend) pathDeleteKeyHandler(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	keyRef := data.Get(keyRefParam).(string)
+	if len(keyRef) == 0 {
+		return logical.ErrorResponse("missing key reference"), nil
+	}
+
+	keyId, err := resolveKeyReference(ctx, req.Storage, keyRef)
+	if err != nil {
+		return nil, err
+	}
+	if keyId == "" {
+		return logical.ErrorResponse("unable to resolve key id for reference" + keyRef), nil
+	}
+
+	keyInUse, issuerId, err := isKeyInUse(keyId.String(), ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if keyInUse {
+		return logical.ErrorResponse(fmt.Sprintf("Failed to Delete Key.  Key in Use by Issuer: %s", issuerId)), nil
+	}
+
+	wasDefault, err := deleteKey(ctx, req.Storage, keyId)
+	if err != nil {
+		return nil, err
+	}
+
+	var response *logical.Response
+	if wasDefault {
+		response = &logical.Response{}
```

Comment:
More changes since you initial created the PR :( 

@cipherboy and I within PR #15178 added a log message on top of the warning to the end user as deleting a default key/issuer would leave us in a bad spot. Would you mind refactoring the message currently within the AddWarning and just output it to a b.Logger().Error(msg) ?

Reply:
Logging an error here makes a lot of sense to me :) done.